Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximated_centroidal_Voronoi_diagram_remeshing.h

```diff
@@ -0,0 +1,1319 @@
+        typename GT::FT k2 = get(vpcd_map, vd).max_curvature;
+        typename GT::FT k_sq = (k1 * k1 + k2 * k2);
+        vertex_weight += weight * pow(k_sq, gradation_factor / 2.0);  // /2.0 because k_sq is squared
+      }
+#endif
+      weight_avg += vertex_weight;
+      put(vertex_weight_pmap, vd, vertex_weight);
+
+      if (use_qem_based_energy)
+      {
+        Matrix4x4 vertex_quadric = get(vertex_quadric_pmap, vd);
+        vertex_quadric += face_quadric;
+        put(vertex_quadric_pmap, vd, vertex_quadric);
+      }
+    }
+  }
+  weight_avg /= nb_vertices;
+
+  // clamp the weights up and below by a ratio (like 10,000) * avg_weights
+  for (Vertex_descriptor vd : vertices(pmesh))
+  {
+    typename GT::FT vertex_weight = get(vertex_weight_pmap, vd);
+    if (vertex_weight > CGAL_WEIGHT_CLAMP_RATIO_THRESHOLD * weight_avg)
+      put(vertex_weight_pmap, vd, CGAL_WEIGHT_CLAMP_RATIO_THRESHOLD * weight_avg);
+    else if (vertex_weight < 1.0 / CGAL_WEIGHT_CLAMP_RATIO_THRESHOLD * weight_avg)
+      put(vertex_weight_pmap, vd, 1.0 / CGAL_WEIGHT_CLAMP_RATIO_THRESHOLD * weight_avg);
+  }
+
+  // randomly initialize clusters
+  //TODO: std::lower_bound with vertex_weight_pmap for better sampling
+  for (int ci = 0; ci < nb_clusters; ++ci)
+  {
+    int vi;
+    Vertex_descriptor vd;
+    do {
+      vi = CGAL::get_default_random().get_int(0, num_vertices(pmesh));
+      vd = *(vertices(pmesh).begin() + vi); // TODO: bad with Polyhedron
+    } while (get(vertex_cluster_pmap, vd) != -1);
+
+    put(vertex_cluster_pmap, vd, ci);
+    typename GT::Point_3 vp = get(vpm, vd);
+    typename GT::Vector_3 vpv(vp.x(), vp.y(), vp.z());
+    clusters[ci].add_vertex(vpv, get(vertex_weight_pmap, vd), get(vertex_quadric_pmap, vd));
+
+    for (Halfedge_descriptor hd : halfedges_around_source(vd, pmesh))
+      clusters_edges_active.push(hd);
+  }
+
+  // the energy minimization loop (clustering loop)
+  int nb_modifications = 0;
+  int nb_disconnected = 0;
+
+  // Turned on once nb_modifications < nb_vertices * CGAL_TO_QEM_MODIFICATION_THRESHOLD
+  bool qem_energy_minimization = false;
+  int nb_loops = 0;
+
+  QEMClusterData<GT> cluster1_v1_to_c2, cluster2_v1_to_c2, cluster1_v2_to_c1, cluster2_v2_to_c1;
+  std::vector<bool> frozen_clusters(nb_clusters, false);
+  do
+  {
+  do
+  {
+    // reset cluster data
+    for (int ci=0; ci<nb_clusters; ++ci)
+      if (!frozen_clusters[ci])
+        clusters[ci] = QEMClusterData<GT>();
+    for ( Vertex_descriptor v : vertices( pmesh ) ) {
+      typename GT::FT v_weight = get(vertex_weight_pmap, v);
+      Matrix4x4 v_qem = get (vertex_quadric_pmap, v);
+      int cluster_id = get(vertex_cluster_pmap, v );
+      if ( cluster_id != -1 && !frozen_clusters[cluster_id])
+        clusters[ cluster_id ].add_vertex( get(vpm, v), v_weight, v_qem);
+    }
+
+     for (int ci=0; ci<nb_clusters; ++ci)
+        if (clusters[ci].nb_vertices==0) throw std::runtime_error("empty_cluster " + std::to_string(ci));
```

Comment:
Do we document that an exception may be thrown, or is there an internal `catch`?

Reply:
this is debugging code I did not have time to clean